include/smack/Naming.h

```diff
@@ -90,6 +91,9 @@ class Naming {
   static const std::string MEMORY_SAFETY_FUNCTION;
   static const std::string MEMORY_LEAK_FUNCTION;
 
+  enum class RMode { RNE, RNA, RTP, RTN, RTZ };
```

Comment:
Ok, I gave this some more thought. The right place for this enum is in the RModeLit class if you think about it. It should not be in Naming.h. I think that we also should not have RMODE_TABLE. I would instead put that directly into the print function of RModeLit.
If you look at where the other naming tables are used, you'll see that they are not used for printing, but instead while particular Boogie expressions are being generated in SmackRep. So I think what I suggested above makes more sense.
Finally, I think if you do what you suggested, you will probably also be able to revert your changes to cmake files since I think you won't run into linking issues.

Reply:
One more thing. See here:
https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly
I think this enum type should be named RModeKind to respect the LLVM coding style.

Reply:
I agree with your reasoning. However, if we go with this, then the Expr class would have the following line:
`static const Expr* rmode(RModeLit::RModeKind v);`
But RModeLit is a subclass of Expr, so this would not be legal (a superclass accessing a subclass's variable). Can you think of a way around this, or am I missing something?

Reply:
Hmmm, right. But enum is just a type and not a class member, i.e., it is not like a class field. So if you make it public in the subclass, can you not simply refer to it in the superclass?
I am not totally sure about this, but it is worth quickly trying it out.

Reply:
It's an issue of circular dependency. For RModeLit::RModeKind to be visible to Expr, RModeLit must be defined before Expr. However, for RModeLit to extend Expr, Expr must be defined before RModeLit. Basically there is no way for Expr to access anything within RModeLit, regardless of whether they are types or class members.